Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,463 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import MobileCoreServices
+
+enum SignalAttachmentError: String {
+    case fileSizeTooLarge
+    case invalidData
+    case couldNotParseImage
+    case couldNotConvertToJpeg
+    case invalidFileFormat
+}
+
+// Represents a possible attachment to upload.
+// The attachment may be invalid.
+//
+// Signal attachments are subject to validation and 
+// in some cases, file format conversion.
+//
+// This class gathers that logic.  It offers factory methods
+// for attachments that do the necessary work. 
+//
+// The return value for the factory methods will be nil if the input is nil.
+//
+// [SignalAttachment hasError] will be true for non-valid attachments.
+//
+// TODO: Perhaps do conversion off the main thread?
+// TODO: Show error on error.
+// TODO: Show progress on upload.
```

Comment:
`SignalAttachment` raises all sorts of interesting questions about how we validate, convert and otherwise handle attachments.  I've tried to capture these questions as TODOs here and below in this file.